sdk/eventhub/Azure.Messaging.EventHubs/samples/Sample11_MockingClientTypes.md

```diff
@@ -0,0 +1,479 @@
+
+bufferedProducerMock
+    .Setup(bp => bp.EnqueueEventAsync(
+        It.IsAny<EventData>(),
+        It.IsAny<CancellationToken>()))
+    .ReturnsAsync(1);
+
+EventHubBufferedProducerClient bufferedProducer = bufferedProducerMock.Object;
+
+// The following demonstrates how to use the mocked EnqueueEventAsync. This would be where application-defined
+// methods can be called and tested.
+
+// This is an illustrative fail handler that can be added to the mock buffered producer.
+
+Func<SendEventBatchFailedEventArgs, Task> sendFailed = new(async args =>
+{
+    foreach (EventData eventData in args.EventBatch)
+    {
+        // Do more complex error handling here...
+
+        Debug.WriteLine($"Publishing failed for {args.EventBatch.Count} events.  Error: '{args.Exception.Message}'");
+        await Task.CompletedTask;
+    }
+});
+
+bufferedProducer.SendEventBatchFailedAsync += sendFailed;
+
+// For illustrative purposes, enqueue a set of events to be published.
+
+int numOfEventsToEnqueue = 15;
+CancellationTokenSource cancellationTokenSource = new();
+
+for (int index = 0; index < numOfEventsToEnqueue; index++)
+{
+    EventData eventToEnqueue = new(index.ToString());
+    await bufferedProducer.EnqueueEventAsync(eventToEnqueue, cancellationTokenSource.Token);
+}
+
+// This shows how to verify that EnqueueEventAsync was called the correct number of times,
+// with the expected arguments.
+
+for (int index = 0; index < numOfEventsToEnqueue; index++)
+{
+    bufferedProducerMock
+        .Verify(bp => bp.EnqueueEventAsync(
+            It.Is<EventData>(e => (e.EventBody.ToString().Equals(index.ToString()))), // Check that enqueue was called on each expected event body
+            It.IsAny<CancellationToken>()));
+}
+
+// The following lines demonstrate how to test a fail handler by calling it directly. Testing fail and success handlers
+// does not require mocking.
+
+int numEventsInBatch = 12;
+List<EventData> eventsInBatch = new();
+
+for (int index = 0; index < numEventsInBatch; index++)
+{
+    EventData eventToEnqueue = new(index.ToString());
+    await bufferedProducer.EnqueueEventAsync(eventToEnqueue, cancellationTokenSource.Token);
+}
+
+SendEventBatchFailedEventArgs args = new SendEventBatchFailedEventArgs(eventsInBatch, new Exception(), "0", default);
+
+// The expected outcome of any application-specific complex processing can be asserted here.
+
+Assert.DoesNotThrowAsync(async () => await sendFailed(args));
+```
+
+## Consuming events using an `EventHubConsumerClient`
+
+When testing code that is dependent on the `EventHubConsumerClient`, a developer may want to create a known set of expected events and contexts to verify that their application consumes them correctly.  Tests can mock `EventData` instances coming from the broker using the `EventHubModelFactory`. These can be used within the consumer client mock to mock returns from `ReadEventsFromPartitionAsync`, which for most applications is the key interaction to test.
```

Comment:
```suggestion
When testing code that is dependent on the `EventHubConsumerClient`, a developer may want to create a known set of expected events to verify that their application consumes them correctly.  Tests can mock `EventData` instances coming from the broker using the `EventHubModelFactory`. These can be used within the consumer client mock to mock returns from `ReadEventsFromPartitionAsync`, which for most applications is the key interaction to test.
```

I think the context is incidental here, the events are the stuff everyone wants.